### PR TITLE
Improving results by multiple supervision steps before grad update

### PR DESCRIPTION
**Pull request by aabbas90**
Adding support for performing multiple backward passes to accumulate gradients from different `N_supervision` steps. This allows to introduce another source of stochasticity in the parameter update. A value of $2$ already seems to give an improvement. (I think $2$ passes are already enough since it can happen that the first pass got halted and the second pass then got a new data sample to compute gradients on). Results on sudoko are as follows

| Experiment name              | Num Training iterations | All.Exact_Accuracy |
|------------------------------|--------------------------|--------------------|
| pretrain_mlp_t_sudoku        | 39060                    | 0.64925            |
| pretrain_mlp_t_sudoku_2_grads| 39060                    | **0.87908**            |

I do not unfortunately have compute resources to check this thoroughly on all datasets. Would be great if anyone can check. Thanks.

### CODE REVIEW

**Review by Happyg76 — APPROVED** [review]

**Review by Happyg76 — CHANGES_REQUESTED** [review]
apologies still needs review

**Comment by olivkoch** [comment]
Trying it on Sudoku Extreme yields a negligible difference (purple = vanilla TRM, green = 2 passes). Can you share your exact config?

Screenshot 2025-12-15 at 16 15 10

**Comment by aabbas90** [comment]
The config is the same as [here](https://github.com/SamsungSAILMontreal/TinyRecursiveModels?tab=readme-ov-file#sudoku-extreme-assuming-1-l40s-gpu) (with MLP). Somehow I was not able to get 87% accuracy with the original code, perhaps I did not setup the requirements correctly. 

image

**Comment by pengzhangzhi** [comment]
> The config is the same as [here](https://github.com/SamsungSAILMontreal/TinyRecursiveModels?tab=readme-ov-file#sudoku-extreme-assuming-1-l40s-gpu) (with MLP). Somehow I was not able to get 87% accuracy with the original code, perhaps I did not setup the requirements correctly.
> 
> image

i wonder if you can reproduce your results consistently?

**Comment by olivkoch** [comment]
I am getting 90% exact eval accuracy on Sudoku Extreme after 20K steps with the official code (batch_size=512) on a single H100. Your two configurations @aabbas90 seem to yield significantly worse performance, so not sure what to conclude from them.
